### PR TITLE
bin-paste: init at 2.0.0

### PR DESCRIPTION
**Pull request by w4**
###### Description of changes

Adds the `bin` package and service, a barebones pastebin that can run with no configuration. The "official" https://bin.gy/ is currently running with this derivation but there has been a couple of requests to move this to Nixpkgs.

https://github.com/w4/bin

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lovesegfault** [comment]
Please call the top-level attr something other than just `bin` if you can? I imagine merging something like this might bite a lot of users that do things like:
```Nix
{ pkgs, ... }:
let
  bin = pkgs.callPackage ./my-bin { };
in
{
  foo = with pkgs; [
    bin
    python
    node
  ];
}
```

File: pkgs/servers/bin/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    rev = "057ed640ee503469fe4c3b0c164aa1b1bffc6cc6";
+    sha256 = "1g67k9m4aflhf2sb2xggh896xgp9vh7n2z0w5kxql0p3xa0shsi9";
+  };
+
+  cargoSha256 = "1v2ks2vdr5knn6kri0d03rvzx1mkv5qhv7xphsxpwfyykhbc0j7h";
+
+  LIBCLANG_PATH="${llvmPackages.libclang}/lib";
+
+  buildInputs = [ ]
+    ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.Security ];
```
Comment (SuperSandro2000) [inline pkgs/servers/bin/default.nix:18]:
```suggestion
  buildInputs = lib.optionals stdenv.isDarwin [ Security ];
```

File: pkgs/servers/bin/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  meta = with lib; {
+    description = "a paste bin.";
```
Comment (SuperSandro2000) [inline pkgs/servers/bin/default.nix:21]:
```suggestion
    description = "Paste bin";
```

File: pkgs/servers/bin/default.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  src = fetchFromGitHub {
+    owner = "w4";
+    repo = "bin";
+    rev = "057ed640ee503469fe4c3b0c164aa1b1bffc6cc6";
+    sha256 = "1g67k9m4aflhf2sb2xggh896xgp9vh7n2z0w5kxql0p3xa0shsi9";
+  };
+
+  cargoSha256 = "1v2ks2vdr5knn6kri0d03rvzx1mkv5qhv7xphsxpwfyykhbc0j7h";
+
+  LIBCLANG_PATH="${llvmPackages.libclang}/lib";
```
Comment (SuperSandro2000) [inline pkgs/servers/bin/default.nix:15]:
```suggestion
  LIBCLANG_PATH = "${llvmPackages.libclang}/lib";
```

File: pkgs/servers/bin/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{ lib, stdenv, rustPlatform, fetchFromGitHub, llvmPackages, darwin }:
+rustPlatform.buildRustPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/servers/bin/default.nix:2]:
```suggestion
{ lib, stdenv, rustPlatform, fetchFromGitHub, llvmPackages, Security }:

rustPlatform.buildRustPackage rec {
```

File: pkgs/servers/bin/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  pname = "bin";
+  version = "2.0.0";
+
+  src = fetchFromGitHub {
+    owner = "w4";
+    repo = "bin";
+    rev = "057ed640ee503469fe4c3b0c164aa1b1bffc6cc6";
```
Comment (SuperSandro2000) [inline pkgs/servers/bin/default.nix:9]:
```suggestion
    rev = "v${version}";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Yeah, we need to find another name. `bin-paste`?

**Comment by w4** [comment]
Good point, @lovesegfault - thanks for spotting that!

Thanks for the review @SuperSandro2000, I've applied your suggestions in [6eb3f99..85a4772](https://github.com/NixOS/nixpkgs/compare/6eb3f9965671a7e54155f461bf406ae7f73cc493..85a47720687aba1f78d07051401653b300712448).

I've renamed the top-level package in nixpkgs to bin-paste, I've left the service as `bin` though, please let me know if that's alright.

Thanks again

**Comment by SuperSandro2000** [comment]
> Edit: I've had to revert [the change from `darwin.apple_sdk.frameworks.Security` to `Security`](https://github.com/NixOS/nixpkgs/pull/165488#discussion_r835904230) as it resulted in errors

You need something similar to the following snippet in all-packages.

```
qtwebview = callPackage ../modules/qtwebview.nix {
  inherit (darwin.apple_sdk.frameworks) CoreFoundation WebKit;
};
```

**Comment by w4** [comment]
Ah, that makes sense! All done, thanks @SuperSandro2000

**Comment by aanderse** [comment]
Sorry for the delay. Modules looks fine. Can you please split your change into 2 commits though? 1 for adding yourself to the maintainer list, and everything else can go into the other commit. Thanks!

**Comment by hacker1024** [comment]
@w4 are you able to split the change?

**Comment by hacker1024** [comment]
@w4?

**Comment by w4** [comment]
Apologies for the delay, this is done now.

File: nixos/modules/services/web-apps/bin-paste.nix [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    package = mkOption {
+      default = pkgs.bin;
+      defaultText = "pkgs.bin-paste";
+      description = "Which bin derivation to use";
+      type = types.package;
+    };
+    maxPasteSize = mkOption {
+      default = 32768;
+      description = "Max allowed size of an individual paste";
+      type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/bin-paste.nix:26]:
```suggestion
      type = types.ints.positive;
```

File: nixos/modules/services/web-apps/bin-paste.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    };
+    bufferSize = mkOption {
+      default = 1000;
+      description = "Maximum amount of pastes to store at a time";
+      type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/bin-paste.nix:31]:
```suggestion
      type = types.ints.positive;
```

File: nixos/modules/services/web-apps/bin-paste.nix [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+in {
+  options.services.bin-paste = {
+    enable = mkEnableOption "bin-paste";
+    bindAddress = mkOption {
+      default = "[::]:8000";
+      description = "Address and port to listen on";
+      type = types.str;
+    };
+    package = mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/bin-paste.nix:17]: Please use mkPackageOption

**Pull request closed without merging**